zarr/core.py

```diff
@@ -1856,11 +1879,59 @@ def _chunk_getitems(self, lchunk_coords, lchunk_selection, out, lout_selection,
                     out[out_select] = fill_value
 
     def _chunk_setitems(self, lchunk_coords, lchunk_selection, values, fields=None):
-        ckeys = [self._chunk_key(co) for co in lchunk_coords]
-        cdatas = [self._process_for_setitem(key, sel, val, fields=fields)
-                  for key, sel, val in zip(ckeys, lchunk_selection, values)]
-        values = {k: v for k, v in zip(ckeys, cdatas)}
-        self.chunk_store.setitems(values)
+        ckeys = map(self._chunk_key, lchunk_coords)
+        cdatas = {key: self._process_for_setitem(key, sel, val, fields=fields)
+                  for key, sel, val in zip(ckeys, lchunk_selection, values)}
+        to_store = {}
+        if not self._write_empty_chunks:
+            empty_chunks = {k: v for k, v in cdatas.items() if self._chunk_is_empty(v)}
+            if hasattr(self.store, 'delitems'):
+                self.store.delitems(tuple(empty_chunks.keys()))
+            else:
+                for ckey in empty_chunks.keys():
+                    self._chunk_delitem(ckey)
+            nonempty_keys = cdatas.keys() - empty_chunks.keys()
+            to_store = {k: self._encode_chunk(cdatas[k]) for k in nonempty_keys}
+        else:
+            to_store = {k: self._encode_chunk(v) for k, v in cdatas.items()}
+        self.chunk_store.setitems(to_store)
+
+    def _chunk_is_empty(self, chunk) -> bool:
```

Comment:
consider adding a code path where we check if the fill value is Falsey, and use np.any, which has some optimizations for this operation.

cc @jakirkham